be/src/exec/pipeline/pipeline_driver_queue.cpp

```diff
@@ -2,80 +2,158 @@
 
-    {
-        std::unique_lock<std::mutex> lock(_global_mutex);
-        while (true) {
-            if (_is_closed) {
-                return Status::Cancelled("Shutdown");
+    for (int i = 0; i < QUEUE_SIZE; ++i) {
+        // we just search for queue has element
+        if (!_queues[i].queue.empty()) {
+            double local_target_time = _queues[i].accu_time_after_divisor();
+            // if this is first queue that has element, we select it;
+            // else we choose queue that the execution time is less sufficient,
+            // and record time.
+            if (queue_idx < 0 || local_target_time < target_accu_time) {
+                target_accu_time = local_target_time;
+                queue_idx = i;
             }
+        }
+    }
 
-            for (int i = 0; i < QUEUE_SIZE; ++i) {
-                // we just search for queue has element
-                if (!_queues[i].queue.empty()) {
-                    double local_target_time = _queues[i].accu_time_after_divisor();
-                    // if this is first queue that has element, we select it;
-                    // else we choose queue that the execution time is less sufficient,
-                    // and record time.
-                    if (queue_idx < 0 || local_target_time < target_accu_time) {
-                        target_accu_time = local_target_time;
-                        queue_idx = i;
-                    }
-                }
-            }
+    DCHECK(queue_idx >= 0);
+
+    driver_ptr = _queues[queue_idx].queue.front();
+    driver_ptr->set_dispatch_queue_index(queue_idx);
+    _queues[queue_idx].queue.pop();
+
+    --_size;
+
+    return driver_ptr;
+}
+
+void QuerySharedDriverQueue::yield_driver(const DriverRawPtr driver) {
+    _queues[driver->get_dispatch_queue_index()].update_accu_time(driver);
+}
+
+void DriverQueueWithWorkGroup::close() {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+    _is_closed = true;
+    _cv.notify_all();
+}
+
+void DriverQueueWithWorkGroup::put_back(const DriverRawPtr driver, bool from_dispatcher) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+    _put_back(driver, from_dispatcher);
+}
 
-            if (queue_idx >= 0) {
-                break;
+void DriverQueueWithWorkGroup::put_back(const std::vector<DriverRawPtr>& drivers, bool from_dispatcher) {
+    std::lock_guard<std::mutex> lock(_global_mutex);
+
+    for (const auto driver : drivers) {
+        _put_back(driver, from_dispatcher);
+    }
+}
+
+StatusOr<DriverRawPtr> DriverQueueWithWorkGroup::take() {
+    std::unique_lock<std::mutex> lock(_global_mutex);
+
+    if (_is_closed) {
+        return Status::Cancelled("Shutdown");
+    }
+
+    while (_wgs.empty()) {
+        _cv.wait(lock);
+        if (_is_closed) {
+            return Status::Cancelled("Shutdown");
+        }
+    }
+
+    auto wg = _find_min_wg();
+    if (wg->driver_queue()->size() == 1) {
+        _sum_cpu_limit -= wg->get_cpu_limit();
+        _wgs.erase(wg);
```

Comment:
Why the _wgs need change? I think most of the time，the _wgs keep immutable

Reply:
`_wgs` only contains the workgroups, which contain ready drivers. Otherwise, there may be a workgroup only containing one ready driver, and multiple dispatcher thread select this workgroup.

Reply:
What's more, the dispatcher thread should block when `_wgs` is empty, so we need remove the workgroup, which doesn't contain the ready drivers.